#[track_caller] in traits

anp opened the pull request:
Per https://github.com/rust-lang/rust/issues/47809#issuecomment-572791760, this allows the `#[track_caller]` attribute on trait methods.

Includes tests for `#[track_caller]` with:

* "regular" trait impls
* default trait impls
* "blanket-tracked" trait impls, where the annotation is in the trait definition and is inherited by "regular" impls of the trait

rust-highfive commented:
r? @sfackler

(rust_highfive has picked a reviewer for you, use r? to override)

anp commented:
cc @eddyb you mentioned wanting to have some particular tests around ReifyShim and the trait impl. I've added traits to the fn-ptr tests, take a look?

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=21221&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21221/logs/70), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21221/logs/70)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-02-18T00:37:55.0028354Z ========================== Starting Command Output ===========================
2020-02-18T00:37:55.0043482Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/633cbf69-9949-47d4-bb8c-c17ecc7bd175.sh
2020-02-18T00:37:55.0200240Z 
2020-02-18T00:37:55.0252962Z ##[section]Finishing: Disable git automatic line ending conversion
2020-02-18T00:37:55.0257509Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/69251/merge to s
2020-02-18T00:37:55.0259210Z Task         : Get sources
2020-02-18T00:37:55.0259237Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-02-18T00:37:55.0259262Z Version      : 1.0.0
2020-02-18T00:37:55.0259287Z Author       : Microsoft
---
2020-02-18T00:37:56.5062420Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-02-18T00:37:56.5147465Z ##[command]git config gc.auto 0
2020-02-18T00:37:56.5206747Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-02-18T00:37:56.5263555Z ##[command]git config --get-all http.proxy
2020-02-18T00:37:56.5402647Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/69251/merge:refs/remotes/pull/69251/merge
---
2020-02-18T01:27:42.4776840Z .................................................................................................... 1700/9648
2020-02-18T01:27:46.8015488Z .................................................................................................... 1800/9648
2020-02-18T01:27:56.9391668Z ..................................i................................................................. 1900/9648
2020-02-18T01:28:03.0754548Z .................................................................................................... 2000/9648
2020-02-18T01:28:14.4136631Z ........................iiiii....................................................................... 2100/9648
2020-02-18T01:28:22.2721405Z .................................................................................................... 2300/9648
2020-02-18T01:28:24.2264148Z .................................................................................................... 2400/9648
2020-02-18T01:28:28.4097446Z .................................................................................................... 2500/9648
2020-02-18T01:28:46.0003759Z .................................................................................................... 2600/9648
---
2020-02-18T01:31:47.4321615Z .................................................................................................... 5600/9648
2020-02-18T01:31:56.2646641Z .......................................................................................i............ 5700/9648
2020-02-18T01:32:02.8384074Z .................................................................................................... 5800/9648
2020-02-18T01:32:07.1940494Z .....................................................................................i.............. 5900/9648
2020-02-18T01:32:15.2343940Z ..............................................................................ii...i..ii...........i 6000/9648
2020-02-18T01:32:33.2746422Z .................................................................................................... 6200/9648
2020-02-18T01:32:38.9020891Z .................................................................................................... 6300/9648
2020-02-18T01:32:44.9551511Z .................................................................................................... 6400/9648
2020-02-18T01:32:44.9551511Z .................................................................................................... 6400/9648
2020-02-18T01:32:56.6937443Z ......i..ii......................................................................................... 6500/9648
2020-02-18T01:33:12.4645479Z ..............................................................................................i..... 6700/9648
2020-02-18T01:33:14.1791060Z .................................................................................................... 6800/9648
2020-02-18T01:33:15.9372525Z .................................................................................................... 6900/9648
2020-02-18T01:33:17.9154491Z .....i.............................................................................................. 7000/9648
---
2020-02-18T01:34:36.9138986Z .................................................................................................... 7600/9648
2020-02-18T01:34:40.7959096Z .................................................................................................... 7700/9648
2020-02-18T01:34:46.1878028Z .................................................................................................... 7800/9648
2020-02-18T01:34:52.0101620Z .................................................................................................... 7900/9648
2020-02-18T01:34:59.7727344Z .....................................................................................iiiiiii.i...... 8000/9648
2020-02-18T01:35:13.5540836Z .........................i......i................................................................... 8200/9648
2020-02-18T01:35:17.8258252Z .................................................................................................... 8300/9648
2020-02-18T01:35:27.8219171Z .................................................................................................... 8400/9648
2020-02-18T01:35:37.6395034Z .................................................................................................... 8500/9648
---
2020-02-18T01:37:40.4454524Z  finished in 6.331
2020-02-18T01:37:40.4608512Z Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-02-18T01:37:40.6165277Z 
2020-02-18T01:37:40.6167543Z running 178 tests
2020-02-18T01:37:44.1569528Z iiii......i...........ii..iiii...i....i...........i............i..i..................i....i......... 100/178
2020-02-18T01:37:45.4934242Z ...i.i.i...iii..iiiiiiiiiiiiiiii.......................iii............ii......
2020-02-18T01:37:45.4936325Z 
2020-02-18T01:37:45.4941408Z  finished in 5.033
2020-02-18T01:37:45.5112135Z Check compiletest suite=codegen-units mode=codegen-units (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-02-18T01:37:45.6528492Z 
---
2020-02-18T01:37:47.6613528Z  finished in 1.832
2020-02-18T01:37:47.6632255Z Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-02-18T01:37:47.6670041Z 
2020-02-18T01:37:47.6670490Z running 9 tests
2020-02-18T01:37:47.6671239Z iiiiiiiii
2020-02-18T01:37:47.6671667Z 
2020-02-18T01:37:47.6671780Z  finished in 0.128
2020-02-18T01:37:47.6672083Z Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-02-18T01:37:47.6672241Z 
---
2020-02-18T01:38:04.4699823Z  finished in 16.968
2020-02-18T01:38:04.4861797Z Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-02-18T01:38:04.6183149Z 
2020-02-18T01:38:04.6183349Z running 116 tests
2020-02-18T01:38:16.3231796Z iiiii..i.....i..i...i..i.i.i..i..i..ii....i.i....ii..........iiii..........i.....i..i.......ii.i.ii. 100/116
2020-02-18T01:38:17.9950705Z ....iiii.....ii.
2020-02-18T01:38:17.9953857Z 
2020-02-18T01:38:17.9958122Z  finished in 13.509
2020-02-18T01:38:17.9962142Z Uplifting stage1 rustc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-02-18T01:38:17.9962628Z Copying stage2 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
---
2020-02-18T01:49:18.8681069Z 
2020-02-18T01:49:18.8685196Z    Doc-tests core
2020-02-18T01:49:22.5692584Z 
2020-02-18T01:49:22.5693261Z running 2471 tests
2020-02-18T01:49:29.8434097Z ......iiiii......................................................................................... 100/2471
2020-02-18T01:49:36.9349289Z ..................................................................................ii................ 200/2471
2020-02-18T01:49:54.3331475Z .................i.................................................................................. 400/2471
2020-02-18T01:49:54.3331475Z .................i.................................................................................. 400/2471
2020-02-18T01:50:01.1334094Z ......................................................................i..i..................iiii.... 500/2471
2020-02-18T01:50:14.8368791Z .................................................................................................... 700/2471
2020-02-18T01:50:21.3592340Z .................................................................................................... 800/2471
2020-02-18T01:50:28.1511478Z .................................................................................................... 900/2471
2020-02-18T01:50:34.9369434Z .................................................................................................... 1000/2471
---
2020-02-18T01:53:21.5331656Z .................................................................................................... 500/760
2020-02-18T01:53:21.5659323Z .................thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libstd/sync/mpsc/mod.rs:2741:22
2020-02-18T01:53:21.5675661Z ....thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libstd/sync/mpsc/mod.rs:2778:21
2020-02-18T01:53:21.5679958Z thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: "SendError(..)"', src/libstd/sync/mpsc/mod.rs:2766:17
2020-02-18T01:53:21.5694469Z .......thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libstd/sync/mpsc/mod.rs:2645:13
2020-02-18T01:53:21.7883631Z ...........................................thread '<unnamed>.' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libstd/sync/mpsc/mod.rs:1997:22
2020-02-18T01:53:21.7902457Z .......thread '<unnamed>...thread '' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libstd/sync/mpsc/mod.rs:2034:21
2020-02-18T01:53:21.7988794Z .................. 600/760
2020-02-18T01:53:23.8222425Z ...................thread '<unnamed>' panicked at 'explicit panic', src/libstd/sync/mutex.rs:633:13
2020-02-18T01:53:23.8227021Z ..thread '<unnamed>' panicked at 'test panic in inner thread to poison mutex', src/libstd/sync/mutex.rs:587:13
2020-02-18T01:53:23.8234980Z ...thread '<unnamed>' panicked at 'test panic in inner thread to poison mutex', src/libstd/sync/mutex.rs:563:13
---
2020-02-18T01:53:32.5990909Z 
2020-02-18T01:53:32.5992472Z running 1009 tests
2020-02-18T01:53:48.7435002Z i................................................................................................... 100/1009
2020-02-18T01:53:57.2387728Z .................................................................................................... 200/1009
2020-02-18T01:54:03.0344327Z ..................iii......i......i...i......i...................................................... 300/1009
2020-02-18T01:54:06.9405416Z .................................................................................................... 400/1009
2020-02-18T01:54:12.3350764Z ............................................i..i.....................................ii............. 500/1009
2020-02-18T01:54:22.2008703Z .................................................................................................... 700/1009
2020-02-18T01:54:22.2008703Z .................................................................................................... 700/1009
2020-02-18T01:54:27.4972588Z ...................................iiii............................................................. 800/1009
2020-02-18T01:54:39.0361572Z .................................................................................................... 900/1009
2020-02-18T01:54:44.3363027Z .........................................................iiii....................................... 1000/1009
2020-02-18T01:54:44.6368517Z test result: ok. 989 passed; 0 failed; 20 ignored; 0 measured; 0 filtered out
2020-02-18T01:54:44.6368799Z 
2020-02-18T01:54:44.6494922Z  finished in 139.280
2020-02-18T01:54:44.6506463Z Testing term stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
2020-02-18T02:11:03.3443966Z   local time: Tue Feb 18 02:11:03 UTC 2020
2020-02-18T02:11:03.6128298Z   network time: Tue, 18 Feb 2020 02:11:03 GMT
2020-02-18T02:11:03.6130096Z == end clock drift check ==
2020-02-18T02:11:03.9947837Z 
2020-02-18T02:11:04.0023793Z ##[error]Bash exited with code '1'.
2020-02-18T02:11:04.0033236Z ##[section]Finishing: Run build
2020-02-18T02:11:04.0055357Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/69251/merge to s
2020-02-18T02:11:04.0057207Z Task         : Get sources
2020-02-18T02:11:04.0057255Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-02-18T02:11:04.0057483Z Version      : 1.0.0
2020-02-18T02:11:04.0057534Z Author       : Microsoft
2020-02-18T02:11:04.0057534Z Author       : Microsoft
2020-02-18T02:11:04.0057741Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
2020-02-18T02:11:04.0057780Z ==============================================================================
2020-02-18T02:11:04.3276231Z Cleaning any cached credential from repository: rust-lang/rust (GitHub)
2020-02-18T02:11:04.3308519Z ##[section]Finishing: Checkout rust-lang/rust@refs/pull/69251/merge to s
2020-02-18T02:11:04.3405866Z Cleaning up task key
2020-02-18T02:11:04.3406471Z Start cleaning up orphan processes.
2020-02-18T02:11:04.3489746Z Terminate orphan process: pid (5519) (python)
2020-02-18T02:11:04.3696754Z ##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

sfackler commented:
r? @eddyb

eddyb commented:
LGTM, as long as we want to do this.

anp commented:
Cool! IIUC the [lang team approved this](https://github.com/rust-lang/rust/issues/47809#issuecomment-572791760).

eddyb commented:
r? @nikomatsakis for the rubber-stamp

eddyb commented on src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs:
```diff
@@ -0,0 +1,16 @@
+
+extern "Rust" {
+    #[track_caller] //~ WARN: `#[track_caller]` is ignored on function prototypes
+    fn bar();
```
Btw we might want this to actually work and change the ABI to pass the caller!

Centril replied:
Would that still apply if the ABI is e.g. `"C"`?

eddyb replied:
I think we'd want that to be an error (it should already be in general, e.g. if you're *defining* an `extern "C" fn`, i.e. exporting it).

Centril replied:
We do have an error without the Rust ABI when exporting:
```rust
#![feature(track_caller)]

#[track_caller]
extern "C" fn foo() {} //~ ERROR `#[track_caller]` requires Rust ABI
```

@anp I don't recall seeing anything about this in the reference PR (https://github.com/rust-lang/reference/pull/742, let's make sure we do note that restriction). Also, do we have a test checking this? (E.g. for "C" and some other random obscure ABI.)

In importing contexts we get `error[E0739]: attribute should be applied to function` right now, which is confusing, but at least there's an error. :)

eddyb replied:
> In importing contexts we get `error[E0739]: attribute should be applied to function` right now, which is confusing, but at least there's an error. :)

This PR changes that to a warning (which is what we're commenting on).

anp replied:
Yeah, let's promote to an error, cover it in the reference, and punt on ABI semantics questions?

Centril replied:
Sounds good. We do have a test `error-with-invalid-abi.rs` :)

anp replied:
https://github.com/rust-lang/reference/pull/742/files#r396035394 for reference docs, changing `warn-fn-prototype.rs` -> `error-extern-fn.rs`.

eddyb commented on src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs:
```diff
@@ -0,0 +1,16 @@
+
+trait Trait {
+    #[track_caller] //~ WARN: `#[track_caller]` is ignored on function prototypes
+    fn foo(&self);
```
We might also want to error here if we want to reserve the right to add meaning to it?

Centril replied:
It does seem prudent to error for now, but I'm curious, do you have thoughts re. meaning?

eddyb replied:
It could be inherited by `impl`s by default, for example.

Centril replied:
That seems plausible, but on the other hand, what is the remedy users would have to disable that effect on specific impls then?

eddyb replied:
They can introduce a non-`#[track_caller]` wrapper, the same way e.g. `libcore` introduces `#[cold]` `#[inline(never)]` functions around panics (although those serve a different purpose).

anp replied:
I'd be very interested in a version of track_caller that's automatically applied to all implementors of a method -- happy to promote warning to error if it creates room for that down the road.

Centril replied:
@anp If you want to follow up with such a PR we could also deal with it before stabilization if you like so that everything is more "complete" when we do stabilize.

anp replied:
Cool. It occurs to me from a recent conversation at work that a blanket track_caller in traits would (I think?) allow error traits to automatically record callchain information with `?`. Exciting if true.

Centril commented:
Let's just confirm that perf doesn't regress; @bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 1e136ee97249970af0f50820b1e3d0dc9b263a98 with merge cb29a8883f4d99b4554b6c987c8e58194efe3d8e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1e136ee97249970af0f50820b1e3d0dc9b263a98","merge_sha":"cb29a8883f4d99b4554b6c987c8e58194efe3d8e"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21276)
Build commit: cb29a8883f4d99b4554b6c987c8e58194efe3d8e (`cb29a8883f4d99b4554b6c987c8e58194efe3d8e`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21276"},"merge_sha":"cb29a8883f4d99b4554b6c987c8e58194efe3d8e"} -->

rust-timer commented:
Queued cb29a8883f4d99b4554b6c987c8e58194efe3d8e with parent b0d5813fd7fe4696b6eb61acf473a7f7a5f9b5a0, future [comparison URL](https://perf.rust-lang.org/compare.html?start=b0d5813fd7fe4696b6eb61acf473a7f7a5f9b5a0&end=cb29a8883f4d99b4554b6c987c8e58194efe3d8e).

anp commented:
I'm not sure what typical variances are on perf.rlo these days, but those numbers don't look great :/.

Centril commented:
I think that's a clear, but not too big regression.

eddyb commented:
You can ignore `syn-opt`, there's an issue open about its nondeterminism.

<hr/>

The weird thing here is `-check` builds being affected by what should be a codegen change only, and especially "clean incremental" runs!

<hr/>

For "baseline incremental" runs the regression is usually `0.3%` or less except for:
* `helloworld-debug` at `0.6%`
* `cranelift-codegen-opt` at `0.7%`
* `token-stream-stress-debug` at `0.4%`

<hr/>

Going back to "clean incremental" runs, the only one over `1.0%` with a change in the number of queries is [`ripgrep-opt`](https://perf.rust-lang.org/detailed-query.html?commit=cb29a8883f4d99b4554b6c987c8e58194efe3d8e&base_commit=b0d5813fd7fe4696b6eb61acf473a7f7a5f9b5a0&benchmark=ripgrep-opt&run_name=clean%20incremental), all the other ones execute the same number of queries, just slower.

<hr/>

Also, many runs that differ in total time, don't differ as much (or at all?) in total query time, that's confusing.

The commonly regressed queries appear to be `incr_comp_load_dep_graph` and... `param_env`?

<hr/>

Hang on, could this entire thing be because of the effect of adding `#[track_caller]` to all of these `impl`s on the compiler code itself?

cc @Mark-Simulacrum any way we could run perf with a `stage1` build, or maybe only enable the `#[cfg_attr(not(bootstrap), track_caller)]` added in this PR for `stage2` `libstd`?

@anp In the meanwhile, maybe we can land the PR without the changed `impl`s? And then maybe figure out separately which ones are responsible for slowing down rustc?

Mark-Simulacrum commented:
> any way we could run perf with a stage1 build, or maybe only enable the #[cfg_attr(not(bootstrap), track_caller)] added in this PR for stage2 libstd?

Stage 1 build will be quite hard I think. The latter, presuming @anp or whoever can rename the attribute to `#[cfg_attr(tracking_caller, track_caller)]` or so, fairly easy, you'd just add some code here gated on stage 2 I guess (or whatever logic you go with): https://github.com/rust-lang/rust/blob/2c462a2f776b899d46743b1b44eda976e846e61d/src/bootstrap/builder.rs#L772

anp commented:
Picking this back up, what do @eddyb and @Centril think about pivoting this PR to implement the attribute-on-trait-prototypes behavior and leaving the specific std annotations for a separate PR, kind of a mix of the suggestions in the subthreads on warn v error and perf runs.

Centril commented:
I think it makes sense to provide the infrastructure first before applying it to specific impls and whatnot.

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=23996&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/23996/logs/71), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/23996/logs/71)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-03-21T17:03:45.2717144Z ========================== Starting Command Output ===========================
2020-03-21T17:03:45.2721483Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/27786f52-feaa-4837-aee0-0bb143236336.sh
2020-03-21T17:03:45.2722002Z 
2020-03-21T17:03:45.2727755Z ##[section]Finishing: Disable git automatic line ending conversion
2020-03-21T17:03:45.2746726Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/69251/merge to s
2020-03-21T17:03:45.2749715Z Task         : Get sources
2020-03-21T17:03:45.2750000Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-03-21T17:03:45.2750295Z Version      : 1.0.0
2020-03-21T17:03:45.2750482Z Author       : Microsoft
---
2020-03-21T17:03:46.2782851Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-03-21T17:03:46.2788078Z ##[command]git config gc.auto 0
2020-03-21T17:03:46.2791356Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-03-21T17:03:46.2794389Z ##[command]git config --get-all http.proxy
2020-03-21T17:03:46.2799788Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/69251/merge:refs/remotes/pull/69251/merge
---
2020-03-21T17:54:29.1038722Z .................................................................................................... 1700/9813
2020-03-21T17:54:32.9232564Z .................................................................................................... 1800/9813
2020-03-21T17:54:42.9270315Z .............................................................................i...................... 1900/9813
2020-03-21T17:54:48.6760755Z .................................................................................................... 2000/9813
2020-03-21T17:54:55.5003498Z ...................................................................iiiii............................ 2100/9813
2020-03-21T17:55:12.3680535Z .................................................................................................... 2300/9813
2020-03-21T17:55:14.3564776Z .................................................................................................... 2400/9813
2020-03-21T17:55:16.9178973Z .................................................................................................... 2500/9813
2020-03-21T17:55:32.3254403Z .................................................................................................... 2600/9813
---
2020-03-21T17:58:06.1335844Z .........................................i...............i.......................................... 5000/9813
2020-03-21T17:58:14.6692447Z .................................................................................................... 5100/9813
2020-03-21T17:58:20.2864943Z .....................................................................................i.............. 5200/9813
2020-03-21T17:58:25.7261790Z .................................................................................................... 5300/9813
2020-03-21T17:58:35.1936326Z ...................................................................ii.ii........i...i............... 5400/9813
2020-03-21T17:58:42.6735254Z .......i............................................................................................ 5600/9813
2020-03-21T17:58:51.3175403Z ............i....................................................................................... 5700/9813
2020-03-21T17:58:56.7031569Z .............................ii...................................i................................. 5800/9813
2020-03-21T17:59:02.9830268Z .................................................................................................... 5900/9813
2020-03-21T17:59:02.9830268Z .................................................................................................... 5900/9813
2020-03-21T17:59:09.0767713Z .................................................................................................... 6000/9813
2020-03-21T17:59:17.4100315Z ............................................................ii...i..ii...........i.................. 6100/9813
2020-03-21T17:59:35.5064917Z .................................................................................................... 6300/9813
2020-03-21T17:59:38.7397476Z .................................................................................................... 6400/9813
2020-03-21T17:59:38.7397476Z .................................................................................................... 6400/9813
2020-03-21T17:59:42.1485308Z ..........................................................................................i..ii..... 6500/9813
2020-03-21T18:00:01.9401480Z .................................................................................................... 6700/9813
2020-03-21T18:00:11.3315250Z .........................................................................................i.......... 6800/9813
2020-03-21T18:00:13.3533792Z .................................................................................................... 6900/9813
2020-03-21T18:00:15.4709707Z .................................................................................................... 7000/9813
---
2020-03-21T18:01:48.8287106Z .................................................................................................... 7800/9813
2020-03-21T18:01:53.3388618Z .................................................................................................... 7900/9813
2020-03-21T18:01:58.9367973Z ...........................................................................i........................ 8000/9813
2020-03-21T18:02:08.0541650Z .................................................................................................... 8100/9813
2020-03-21T18:02:12.9786569Z ........................iiiiiiiiii.i................................................................ 8200/9813
2020-03-21T18:02:25.7285215Z .................................................................................................... 8400/9813
2020-03-21T18:02:30.6770017Z .................................................................................................... 8500/9813
2020-03-21T18:02:44.5105070Z .................................................................................................... 8600/9813
2020-03-21T18:02:50.9090939Z .................................................................................................... 8700/9813
---
2020-03-21T18:04:57.4349142Z Suite("src/test/codegen") not skipped for "bootstrap::test::Codegen" -- not in ["src/tools/tidy"]
2020-03-21T18:04:57.4532186Z Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-03-21T18:04:57.6349876Z 
2020-03-21T18:04:57.6350223Z running 183 tests
2020-03-21T18:05:00.1093426Z iiii......i............ii.i..iiii....i....i...........i............i..i..................i....i..... 100/183
2020-03-21T18:05:02.4268589Z .......i.i.i...iii..iiiiiiiiiiiiiiii.......................iii.............ii......
2020-03-21T18:05:02.4275824Z 
2020-03-21T18:05:02.4280371Z  finished in 4.974
2020-03-21T18:05:02.4284334Z Suite("src/test/codegen-units") not skipped for "bootstrap::test::CodegenUnits" -- not in ["src/tools/tidy"]
2020-03-21T18:05:02.4460593Z Check compiletest suite=codegen-units mode=codegen-units (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
2020-03-21T18:05:04.3292664Z Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
2020-03-21T18:05:04.3481749Z Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-03-21T18:05:04.4935798Z 
2020-03-21T18:05:04.4936693Z running 9 tests
2020-03-21T18:05:04.4937727Z iiiiiiiii
2020-03-21T18:05:04.4938815Z 
2020-03-21T18:05:04.4938959Z  finished in 0.145
2020-03-21T18:05:04.4939528Z Suite("src/test/incremental") not skipped for "bootstrap::test::Incremental" -- not in ["src/tools/tidy"]
2020-03-21T18:05:04.5116344Z Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
2020-03-21T18:05:22.5169405Z Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
2020-03-21T18:05:22.5376507Z Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-03-21T18:05:22.7023514Z 
2020-03-21T18:05:22.7024071Z running 115 tests
2020-03-21T18:05:34.7351182Z iiiii..i.....i..i...i..i.i.i..i..i..ii....i.i....ii..........iiii.........i.....i..i.......ii.i.ii.. 100/115
2020-03-21T18:05:36.2326659Z ...iiii.....ii.
2020-03-21T18:05:36.2327870Z 
2020-03-21T18:05:36.2330865Z  finished in 13.694
2020-03-21T18:05:36.2331438Z Uplifting stage1 rustc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-03-21T18:05:36.2332081Z Copying stage2 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
---
2020-03-21T18:16:41.7812105Z 
2020-03-21T18:16:41.7813555Z    Doc-tests core
2020-03-21T18:16:45.8094627Z 
2020-03-21T18:16:45.8095314Z running 2484 tests
2020-03-21T18:16:53.9588396Z ......iiiii......................................................................................... 100/2484
2020-03-21T18:17:01.9044584Z .....................................................................................ii............. 200/2484
2020-03-21T18:17:20.5427521Z ....................i............................................................................... 400/2484
2020-03-21T18:17:20.5427521Z ....................i............................................................................... 400/2484
2020-03-21T18:17:29.4994348Z .........................................................................i..i..................iiii. 500/2484
2020-03-21T18:17:44.5626299Z .................................................................................................... 700/2484
2020-03-21T18:17:52.3805879Z .................................................................................................... 800/2484
2020-03-21T18:17:59.8619877Z .................................................................................................... 900/2484
2020-03-21T18:18:07.6417033Z .................................................................................................... 1000/2484
---
2020-03-21T18:21:17.8661944Z 
2020-03-21T18:21:17.8664062Z running 1012 tests
2020-03-21T18:21:33.1432373Z i................................................................................................... 100/1012
2020-03-21T18:21:42.0263282Z .................................................................................................... 200/1012
2020-03-21T18:21:48.2678578Z ..................iii......i......i...i......i...................................................... 300/1012
2020-03-21T18:21:58.5620747Z ............................................i....i......................................ii.......... 500/1012
2020-03-21T18:22:04.9263062Z .................................................................................................... 600/1012
2020-03-21T18:22:09.5672357Z .................................................................................................... 700/1012
2020-03-21T18:22:09.5672357Z .................................................................................................... 700/1012
2020-03-21T18:22:15.4864692Z ......................................iiii.......................................................... 800/1012
2020-03-21T18:22:27.7187983Z .................................................................................................... 900/1012
2020-03-21T18:22:33.5648689Z ............................................................iiii.................................... 1000/1012
2020-03-21T18:22:33.9949101Z test result: ok. 992 passed; 0 failed; 20 ignored; 0 measured; 0 filtered out
2020-03-21T18:22:33.9949351Z 
2020-03-21T18:22:34.0051051Z  finished in 145.316
2020-03-21T18:22:34.0052300Z Set({"src/libterm"}) not skipped for "bootstrap::test::Crate" -- not in ["src/tools/tidy"]
---
2020-03-21T18:39:28.9617228Z   local time: Sat Mar 21 18:39:28 UTC 2020
2020-03-21T18:39:29.1852915Z   network time: Sat, 21 Mar 2020 18:39:29 GMT
2020-03-21T18:39:29.1856895Z == end clock drift check ==
2020-03-21T18:39:29.5615922Z 
2020-03-21T18:39:29.5680079Z ##[error]Bash exited with code '1'.
2020-03-21T18:39:29.5704574Z ##[section]Finishing: Run build
2020-03-21T18:39:29.5760445Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/69251/merge to s
2020-03-21T18:39:29.5765439Z Task         : Get sources
2020-03-21T18:39:29.5765763Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-03-21T18:39:29.5766077Z Version      : 1.0.0
2020-03-21T18:39:29.5766289Z Author       : Microsoft
2020-03-21T18:39:29.5766289Z Author       : Microsoft
2020-03-21T18:39:29.5766618Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
2020-03-21T18:39:29.5767021Z ==============================================================================
2020-03-21T18:39:29.8797740Z Cleaning any cached credential from repository: rust-lang/rust (GitHub)
2020-03-21T18:39:29.8852203Z ##[section]Finishing: Checkout rust-lang/rust@refs/pull/69251/merge to s
2020-03-21T18:39:29.8937580Z Cleaning up task key
2020-03-21T18:39:29.8939048Z Start cleaning up orphan processes.
2020-03-21T18:39:29.9109523Z Terminate orphan process: pid (3596) (python)
2020-03-21T18:39:29.9335176Z ##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Centril commented:
Reassigning to technical reviewer as lang team has given its blessing. r? @eddyb

anp commented:
I split the perf-sensitive change out in https://github.com/rust-lang/rust/pull/70234.

Added tests here for the use of the attribute on trait method prototypes taking effect on impls (as discussed in review thread), now need to make it work.

rust-highfive commented:
The job `mingw-check` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=24044&view=logs&jobId=74054d28-c774-5fab-ecc2-352fe71b230e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/24044/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/24044/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-03-22T02:30:47.0888657Z ========================== Starting Command Output ===========================
2020-03-22T02:30:47.0890962Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/534a5675-cc8c-4b89-8330-2029162e4e93.sh
2020-03-22T02:30:47.0891206Z 
2020-03-22T02:30:47.0894355Z ##[section]Finishing: Disable git automatic line ending conversion
2020-03-22T02:30:47.0915047Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/69251/merge to s
2020-03-22T02:30:47.0917934Z Task         : Get sources
2020-03-22T02:30:47.0918200Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-03-22T02:30:47.0918443Z Version      : 1.0.0
2020-03-22T02:30:47.0918606Z Author       : Microsoft
---
2020-03-22T02:30:48.4245502Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-03-22T02:30:48.4251525Z ##[command]git config gc.auto 0
2020-03-22T02:30:48.4256312Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-03-22T02:30:48.4260460Z ##[command]git config --get-all http.proxy
2020-03-22T02:30:48.4269159Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/69251/merge:refs/remotes/pull/69251/merge
---
2020-03-22T02:39:15.3185076Z configure: build.locked-deps    := True
2020-03-22T02:39:15.3185347Z configure: llvm.ccache          := sccache
2020-03-22T02:39:15.3185784Z configure: build.cargo-native-static := True
2020-03-22T02:39:15.3186230Z configure: dist.missing-tools   := True
2020-03-22T02:39:15.3186769Z configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
2020-03-22T02:39:15.3187298Z configure: writing `config.toml` in current directory
2020-03-22T02:39:15.3187517Z configure: 
2020-03-22T02:39:15.3187887Z configure: run `python /checkout/x.py --help`
2020-03-22T02:39:15.3188112Z configure: 
---
2020-03-22T02:45:31.6435114Z    Compiling cargo_metadata v0.9.1
2020-03-22T02:45:34.7122755Z    Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
2020-03-22T02:45:44.1463196Z     Finished release [optimized] target(s) in 21.29s
2020-03-22T02:45:44.1570921Z tidy check
2020-03-22T02:45:50.3996818Z tidy error: /checkout/src/librustc_error_codes/error_codes/E0738.md:1: line longer than 80 chars
2020-03-22T02:45:53.0198940Z Found 489 error codes
2020-03-22T02:45:53.0201190Z Found 0 error codes with no tests
2020-03-22T02:45:53.0201767Z Done!
2020-03-22T02:45:53.0202115Z some tidy checks failed
2020-03-22T02:45:53.0202115Z some tidy checks failed
2020-03-22T02:45:53.0205388Z 
2020-03-22T02:45:53.0205640Z 
2020-03-22T02:45:53.0206848Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo"
2020-03-22T02:45:53.0207557Z 
2020-03-22T02:45:53.0207649Z 
2020-03-22T02:45:53.0220978Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
2020-03-22T02:45:53.0221520Z Build completed unsuccessfully in 0:00:31
2020-03-22T02:45:53.0221520Z Build completed unsuccessfully in 0:00:31
2020-03-22T02:45:53.0272492Z == clock drift check ==
2020-03-22T02:45:53.0291483Z   local time: Sun Mar 22 02:45:53 UTC 2020
2020-03-22T02:45:53.1221396Z   network time: Sun, 22 Mar 2020 02:45:53 GMT
2020-03-22T02:45:53.1225436Z == end clock drift check ==
2020-03-22T02:45:54.8262940Z 
2020-03-22T02:45:54.8302501Z ##[error]Bash exited with code '1'.
2020-03-22T02:45:54.8313606Z ##[section]Finishing: Run build
2020-03-22T02:45:54.8364106Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/69251/merge to s
2020-03-22T02:45:54.8368812Z Task         : Get sources
2020-03-22T02:45:54.8369149Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-03-22T02:45:54.8369467Z Version      : 1.0.0
2020-03-22T02:45:54.8369676Z Author       : Microsoft
2020-03-22T02:45:54.8369676Z Author       : Microsoft
2020-03-22T02:45:54.8370117Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
2020-03-22T02:45:54.8370451Z ==============================================================================
2020-03-22T02:45:55.1202093Z Cleaning any cached credential from repository: rust-lang/rust (GitHub)
2020-03-22T02:45:55.1261606Z ##[section]Finishing: Checkout rust-lang/rust@refs/pull/69251/merge to s
2020-03-22T02:45:55.1340723Z Cleaning up task key
2020-03-22T02:45:55.1341904Z Start cleaning up orphan processes.
2020-03-22T02:45:55.1609782Z Terminate orphan process: pid (3891) (python)
2020-03-22T02:45:55.1650456Z ##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -2583,6 +2586,34 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
     codegen_fn_attrs
 }
 
+/// Checks if the provided DefId implements a method for a trait which has track_caller applied to the trait prototype.
```
This shouldn't pass tidy.

Also, regarding the contents themselves, "implements a method" could be expanded to describe the fact that this is method in a trait impl.

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -2583,6 +2586,35 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
+                    .associated_items(trait_id)
+                    .filter_by_name_unhygienic(assoc.ident.name)
+                    .find(move |impl_item| match impl_item.kind {
+                        ty::AssocKind::Method => {
```
`impl_item.kind == ty::AssocKind::Method && ...`

eddyb commented on src/librustc/ty/mod.rs:
```diff
@@ -2875,8 +2875,8 @@ impl<'tcx> TyCtxt<'tcx> {
-            match self.def_kind(def_id).expect("no def for `DefId`") {
-                DefKind::AssocConst | DefKind::AssocFn | DefKind::AssocTy => true,
+            match self.def_kind(def_id) {
+                Some(DefKind::AssocConst | DefKind::AssocFn | DefKind::AssocTy) => true,
```
Now that this is fixed, maybe this special case can be removed: https://github.com/rust-lang/rust/blob/d1e81ef234ff5c2e0e3a69cb4e8e5f5b0fe1fd83/src/librustc/ty/mod.rs#L3057-L3063

eddyb commented on src/librustc_passes/check_attr.rs:
```diff
@@ -151,17 +151,17 @@ impl CheckAttrVisitor<'tcx> {
                     *attr_span,
                     E0738,
-                    "`#[track_caller]` may not be used on trait methods",
+                    "`#[track_caller]` is not supported on imported function prototypes",
```
Is this language used elsewhere? I would've expected "on foreign functions".

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -2583,6 +2586,35 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
+fn should_inherit_track_caller(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
+    if let Some(assoc) = tcx.opt_associated_item(def_id) {
+        if let ty::AssocItemContainer::ImplContainer(container) = assoc.container {
+            if let Some(trait_id) = tcx.trait_id_of_impl(container) {
```
`container` -> `impl_def_id`

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -2583,6 +2586,35 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
+                        }
+                        _ => false,
+                    })
+                    .copied()
```
Shouldn't need `.copied()`, AFAICT the element type is a `&'tcx ty::AssocItem`.

anp replied:
it's my fault bc i `.copied()` from the example query in the specialization graph you showed me!

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -2583,6 +2586,35 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
+                if let Some(trait_method_id) = tcx
+                    .associated_items(trait_id)
+                    .filter_by_name_unhygienic(assoc.ident.name)
+                    .find(move |impl_item| match impl_item.kind {
```
`impl_item` -> `trait_item`

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -2583,6 +2586,35 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
+    if let Some(assoc) = tcx.opt_associated_item(def_id) {
+        if let ty::AssocItemContainer::ImplContainer(container) = assoc.container {
+            if let Some(trait_id) = tcx.trait_id_of_impl(container) {
+                if let Some(trait_method_id) = tcx
```
`trait_method_id` -> `trait_item`

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -2583,6 +2586,35 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
+/// Checks if the provided DefId is a method in a trait impl for a trait which has track_caller 
+/// applied to the method prototype.
+fn should_inherit_track_caller(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
+    if let Some(assoc) = tcx.opt_associated_item(def_id) {
```
`assoc` -> `impl_item` (unless you want to use some suffix other than `_item`)

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -2583,6 +2586,32 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
+fn should_inherit_track_caller(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
+    if let Some(impl_item) = tcx.opt_associated_item(def_id) {
+        if let ty::AssocItemContainer::ImplContainer(impl_def_id) = impl_item.container {
+            if let Some(trait_id) = tcx.trait_id_of_impl(impl_def_id) {
```
`trait_id` -> `trait_def_id`

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -2583,6 +2586,32 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
+                    .filter_by_name_unhygienic(impl_item.ident.name)
+                    .find(move |trait_item| {
+                        trait_item.kind == ty::AssocKind::Method
+                            && tcx.hygienic_eq(trait_item.ident, impl_item.ident, trait_id)
```
Order is `use_ident`, `def_ident`, `def_parent`.
So `impl_item.ident`, `trait_item.ident`, `trait_def_id`.

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -2583,6 +2586,32 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
+                    return tcx
+                        .codegen_fn_attrs(trait_item.def_id)
+                        .flags
+                        .intersects(CodegenFnAttrFlags::TRACK_CALLER);
```
I wonder if we should iterate over specialization ancestors, so that each of them can add `#[track_caller]`, not just the `trait` and the `fn` with a body.

anp replied:
That seems like the "most correct" option, unsure the implementation complexity is worth it. If you think it'll be easy can you give me some pointers? If not, I'll file a bug?

eddyb replied:
Better to file an issue. Specialization is unstable anyway.

anp replied:
https://github.com/rust-lang/rust/issues/70293

eddyb commented on src/test/ui/rfc-2091-track-caller/tracked-trait-impls.rs:
```diff
@@ -0,0 +1,40 @@
+    #[track_caller]
+    fn default_tracked(&self) {
+        check_caller_is_tracked(line!());
+    }
```
Can you add a second `impl` where you override this default method?

anp replied:
There are two possible cases when overriding the default: with the attribute and without. Should both have the codegen flag?

anp replied:
I'm going with "yes" since that's what happens by default with the codegen flag inheritance.

eddyb commented on src/test/ui/rfc-2091-track-caller/tracked-trait-impls.rs:
```diff
@@ -0,0 +1,40 @@
+    let location = std::panic::Location::caller();
+    assert_eq!(location.file(), file!());
+    assert_ne!(location.line(), line_called_at, "line should be outside this fn");
+}
```
This sort of thing could be a macro, but I don't care too much either way.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 6c2c559dab1c3a353f9cc57501a4d155baadfdf2 has been approved by `eddyb`

<!-- @bors r=eddyb 6c2c559dab1c3a353f9cc57501a4d155baadfdf2 -->
<!-- homu: {"type":"Approved","sha":"6c2c559dab1c3a353f9cc57501a4d155baadfdf2","approver":"eddyb"} -->

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 69bd46a6a18a5b44972ccebb96de1294f4b760b1 has been approved by `eddyb`

<!-- @bors r=eddyb 69bd46a6a18a5b44972ccebb96de1294f4b760b1 -->
<!-- homu: {"type":"Approved","sha":"69bd46a6a18a5b44972ccebb96de1294f4b760b1","approver":"eddyb"} -->
